---
title: "Accessing Data Sets"
output:
  BiocStyle::html_document:
    toc: true
    number_section: true
    self_contained: true
    titlecaps: true
vignette: >
  %\VignetteIndexEntry{Accessing Data Sets}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, include = FALSE}
library(scMultiome)
```

# Available Data Sets

Use the `listDatasets` function to view currently available data sets and the names of their accessor functions. Help pages fot particular accessors contain more information on their data sets.
```{r, eval = FALSE}
library(scMultiome)
listDatasets()
```
```{r, echo = FALSE, results = 'asis'}
lds <- listDatasets()
knitr::kable(lds, caption = "Available Data Sets")
```

Access a data set by calling its accessor function:
```{r eval = FALSE}
prostateENZ()
```


See the help files for accessor functions for possible extra options, e.g. `?prostateENZ`.



<br><br><br>

# Transcription Factor Binding Motifs

The package contains additional data sets that contain information on transcription factor binding. These are not multiome data but compiled ChIP-seq data that can be used together with multiomes to compute epiregulons. Currently three genomic builds genomes are provided: hg38, hg19, and mm10. Each genome was built with combined ChIP-seq data from ChIP-Atlas and ENCODE or CistromeDB and ENCODE.

The ChIP-seq data is packaged into individual RDS files but they are accessed with a common accessor function, `tfBinding`. Only one build can be retrieved at a time, a genome and data source must be specified.
```{r, eval = FALSE}
tfBinding("hg38", "atlas")
```



<br><br><br>
