---
title: "Multiome Data Processing"
output:
  BiocStyle::html_document:
    toc: true
    number_section: true
    self_contained: true
    titlecaps: true
vignette: >
  %\VignetteIndexEntry{Obtaining Multiome Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# Introduction

A multiome typically consists of chromatin accessibility (scATAC) and gene expression (scRNAseq) data subsets, either paired/multimodal (both assays were run on material extracted from the same set of samples and as such they contain the same cell barcodes) or unpaired (material extraction and barcoding were performed independently).

The multiome data herein comes from various sources but were all analyzed with the `ArchR` package.



<br><br><br>

# Raw Data Analysis

Raw data is first analyzed with the [10x Genomics Cell Ranger](https://www.10xgenomics.com/support) pipelines.

Following that, gene expression data is run through a [scran.chan](https://github.com/LTLA/scran.chan) analysis pipeline. Refer to the [OSCA book](https://bioconductor.org/books/release/OSCA/), specifically [this chapter](http://bioconductor.org/books/3.15/OSCA.basic/index.html) for details.



<br><br><br>

# ArchR Multiome Workflow

The general ArchR workflow is outlined below.
Refer to the [ArchR book](https://www.archrproject.com/bookdown/index.html) for details.

1. Fragment files resulting from Cell Ranger ATAC analysis are used to create arrow files. The _Tile Matrix_ and _Gene Score Matrix_ are created.
2. An ArchR project is created. Project-wide Tile and Gene Score Matrices are now available.
3. Duplex cells are filtered out. This step is omitted in cell culture based experiments.
4. Dimensionality reduction is run on the Tile Matrix using the Iterative LSI algorithm using `ArchR::addIterativeLSI`.
5. scRNAseq data is added to the ArchR project with `ArchR::addGeneExpressionMatrix` for paired datasets or with `ArchR::addGeneIntegrationMatrix` for unpaired ones. The _Gene Expression Matrix_ or the _Gene Integration Matrix_ is added.
6. Iterative LSI is run again on the _Gene Expression/Integration Matrix_.
7. Reduced dimension representations of scATAC and scRNAseq spaces are combined with `ArchR::addCombinedDims`. This creates a joint reduced dimensions representation that is used downstream.
8. Clustering is done with `ArchR::addClusters` on the iLSI representation of a space of the user's choosing: the scATAC space, the scRNAseq space, or the combined space..
9. UMAP and/or TSNE embeddings are created with `ArchR::addUMAP` and `ArchR::addTSNE`. Likewise, the user chooses the iLSI space to create embeddings from.
10. Pseudo-bulk replicates are computed with `ArchR::addGroupCoverages`.
11. Peaks are called with `ArchR::addReproduciblePeakSet`. (This requires `MACS2`.)
12. _Peak Matrix_ is added with `ArchR::addPeakMatrix`.
13. Peak annotation is added with `ArchR::addMotifAnnotations`. (This requires `chromVARmotifs`.)
14. Background peaks are obtained with `ArchR::addBgdPeaks`.
15. ChromVAR deviations are computed with `ArchR::addDeviationsMatrix`. The _Motif Matrix_ is added.



<br><br><br>

# Conversion to MAE

ArchR projects were converted to `MultiAssayExperiment` objects with `archr2MAE`. Each Matrix in the project becomes a `SingleCellExperiment`, its `colData` is derived from the project's `cellColData` slot, and `rowData` and `rowRanges` are created as needed. Reduced dimensionalities and embeddings of the project are attached to the Matrices they were obtained from and are stored in the `reducedDims` slot.



<br><br><br>

# Storage And Access

The ArchR MAE is stored in a hdf5 file. In order to reduce storage space, assays are extracted from Matrices and saved in a compressed format (see `?writeSparseMatrix`). This requires that metadata is stored separately from the assays (but still in the same file). The data is accessed by functions that extract requested assays and their respective metadata, reconstruct the SCE(s) and return an MAE.



<br><br><br>

# Contributing

If you want to contribute your publicly available multiome data set, please read the vignette _Adding Data Sets_ and contact the package maintainer, `r desc::desc_get_maintainer()`.



<br><br><br>
