---
title: "Developer Notes"
output:
    BiocStyle::html_document:
    toc: true
number_section: true
self_contained: true
titlecaps: true
vignette: >
  %\VignetteIndexEntry{Developer Notes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{css, echo=FALSE}
ol {
  padding: 0.25em 1.5em;
  margin: 0.5em 0em 0em 0em;
}
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

# Package Creation History

This is how the package was created:

1. Initiate a package project with RStudio.

2. Clean up unwanted files.
```{r}
unlink(list.files("R", full.names = TRUE))
unlink(list.files("man", full.names = TRUE))
unlink("NAMESPACE")
```

3. Add key files.
```{r}
dir.create("inst/scripts", recursive = TRUE)
dir.create("inst/extdata", recursive = TRUE)
file.create("./inst/scripts/make-data.R")
file.create("./inst/scripts/make-metadata.R")
```

4. Modify DESCRIPTION file.
```{r}
# delete fields
desc::desc_del(c("Author", "Maintainer"))
# add authors with desc::desc_add_author
# (always add one author and one role at a time)
desc::desc_add_author("Xiaosai", "Yao", "xyao19@gene.com", "cre")
desc::desc_add_author("Aleksander", "Chlebowski", "aleksander.chlebowski@contractors.roche.com", "aut")
# edit fields
desc::desc_set("Package", "scMultiome")
desc::desc_set("Title", "Collection of Public Single-Cell Multiome (scATAC + scRNAseq) Datasets")
desc::desc_set("Description", "Single cell multiome data, containing chromatin accessibility (scATAC-seq) and gene expression (scRNA-seq) information analyzed with the ArchR package and presented as MultiAssayExperiment objects.")
desc::desc_set("License", "Artistic-2.0")
# add biocViews field (required by Bioconductor)
desc::desc_set("biocViews", "ExperimentHub, SingleCellData, ExpressionData, Homo_sapiens_Data, CellCulture, Tissue, GEO")
# enable Rmarkdown support for help files
# this allows inserting unevaluated code chunks with ```R and quoting Rmd files with {r child = <PATH>}
desc::desc_set("Roxygen", "list(markdown = TRUE)")
# add dependencies
usethis::use_package("SummarizedExperiment", type = "Depends") # all experiments inherit from this
usethis::use_package("SingleCellExperiment", type = "Depends") # all experiments inherit from this
usethis::use_package("MultiAssayExperiment", type = "Depends") # class of final object
usethis::use_package("AnnotationHub", type = "Depends")        # required by Bioconductor
usethis::use_package("ExperimentHub", type = "Depends")        # required by Bioconductor
usethis::use_package("checkmate")                              # for argument checks in functions
usethis::use_package("S4Vectors")                              # for operating on DataFrames
usethis::use_package("GenomicRanges")                          # for manipulating GRanges
usethis::use_package("rhdf5")                                  # for operating on hdf5 files
usethis::use_package("methods")                                # for testing and changing object classes
usethis::use_package("HDF5Array")                              # for reading assays from hdf5 files
usethis::use_package("DelayedArray")                           # loaded assays inherit from this
usethis::use_package("chihaya")                                # for writing matrices
usethis::use_package("AzureStor", type = "Suggests")           # for uploading data to Bioconductor
usethis::use_package("BiocGenerics", type = "Suggests")        # for unit tests
usethis::use_package("IRanges", type = "Suggests")             # for unit tests
usethis::use_package("Matrix", type = "Suggests")              # for unit tests
usethis::use_package("rstudioapi", type = "Suggests")          # for template functions
# set development version
desc::desc_set_version("0.0.0.9000")
# normalize file
desc::desc_normalize()
```

5. Add package documentation.
```{r}
usethis::use_package_doc()
```

6. Build man files.
```{r}
devtools::document()
```

7. Create vignette.
```{r}
usethis::use_vignette(name = "multiome", title = "Obtaining Multiome Data")
```



<br><br><br>

# Notes for the Future

## Formatting Rmd files

Use this output specification in Rmd files, including vignette(s):
```
output:
  BiocStyle::html_document:
  toc: true
number_section: true
self_contained: true
titlecaps: true
```

<br>

## Code evaluation in Rmd files

Disable evaluation of all code chunks in an RMD file:
```
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

Disable evaluation for a specific chunk (omit quote marks):
````{verbatim}
```{r, eval = FALSE}

```
````

<br>

## Suggested packages

Using dependencies listed as "Suggests" should be gated by `if (requireNamespace("<PACKAGE_NAME>", quitely = TRUE))` statements.

Having dependencies not installed may be seen as a problem by `R CMD CHECK`. You may have to set `force_suggests = FALSE` when running `devtools::check()`.

<br>

## Vignettes

If a vignette contains code that accesses a data set that has not been placed in `ExperimentHub` yet, it will throw an error dirung `R CMD CHECK`. Add a `eval = FALSE` when checking the package.

<br>
