---
title: "Obtaining Multiome Data"
output:
  BiocStyle::html_document:
    toc: true
    number_section: true
    self_contained: true
    titlecaps: true
vignette: >
  %\VignetteIndexEntry{Obtaining Multiome Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


# Introduction

A multiome typically consists of chromatin accessibility (scATAC) and gene expression (scRNAseq) data subsets, either paired/multimodal (both assays were run on material extracted from the same set of samples and as such they contain the same cell barcodes) or unpaired (material extraction and barcoding were performed independently).

The multiome data herein comes from various sources. The general workflow is as follows:

1. Raw data is inducted to projects in Genentech's FireDB, a structured database of resources.
2. FireDB resources are sent for analysis to Genentech's Sunrise genomic analysis portal.
3. Sunrise runs the 10x Genomics Cell Ranger pipelines appropriate for particular datasets.
4. Analysis results are added to the respective FireDB projects and stored on a Genentech storage server. At this stage scRNAseq data are also added to DataSetDB, which is an internal database.
5. Analysis results are imported into an R session and downstream analysis is performed accordingly.
6. During the analysis the scATAC and scRNAseq data are combined.



<br><br><br>

# Raw Data Analysis

Raw data is first analyzed with the [10x Genomics Cell Ranger](https://www.10xgenomics.com/support) pipelines.

Following that, the gene expression data set from DataSetDB is passed through the `gp.sa.solo::runSingleCellBasic` or `gp.sa.solo::runSingleCellMerge` for quality control and, in case of the latter, batch effect correction. Refer to the [OSCA book](https://bioconductor.org/books/release/OSCA/), specifically [this chapter](http://bioconductor.org/books/3.15/OSCA.basic/index.html) for details.



<br><br><br>

# ArchR Multiome Workflow

The general ArchR workflow rus outlined below.
Refer to the [ArchR book](https://www.archrproject.com/bookdown/index.html) for details.

1. Fragment files resulting from Cell Ranger ATAC analysis are used to create arrow files.
2. An ArchR project is created. The Tile Matrix and Gene Score Matrix are created.
3. Duplex cells are filtered out. This step is omitted in cell culture based experiments.
4. Dimensionality reduction is run on the Tile Matrix using the Iterative LSI algorithm using `ArchR::addIterativeLSI`.
5. scRNAseq data is added to the ArchR project with `ArchR::addGeneExpressionMatrix` for paired datasets or with `ArchR::addGeneIntegrationMatrix` for unpaired ones. The Gene Expression Matrix or the Gene Integration Matrix is added.
6. Iterative LSI is run again on the Gene Expression/Integration Matrix.
7. Reduced dimension representations of scATAC and scRNAseq spaces are combined with `ArchR::addCombinedDims`. This creates a joint reduced dimensions representation that is used downstream.
8. Clustering is done with `ArchR::addClusters`.
9. UMAP and/or TSNE embeddings are created with `ArchR::addUMAP` and `ArchR::addTSNE`.
10. Pseudo-bulk replicates are computed with `ArchR::addGroupCoverages`. (This requires `MACS2`.)
11. Peaks are called with `ArchR::addReproduciblePeakSet`. (This requires `chromVARmotifs`.)
12. Peak Matrix is added with `ArchR::addPeakMatrix`.
13. Peak annotation is added with `ArchR::addMotifAnnotations`.
14. Background peaks are obtained with `ArchR::addBgdPeaks`.
15. ChromVAR deviations are computed with `ArchR::addDeviationsMatrix`. The Motif Matrix is added.



<br><br><br>

# Conversion to MAE

ArchR projects were converted to `MultiAssayExperiment` objects using the function `maw.archr::create.mae.with.multiple.sces.from.archr`. Each Matrix in the project becomes a `SingleCellExperiment` object in the resulting MAE.



<br><br><br>

# Storage And Access

The ArchR MAE is stored in a hdf5 file. In order to reduce storage space, assays are extracted from Matrices and saved with `artificer.matrix::writeSparseMatrix`. This requires that metadata is stored separately from the assays (but still in the same file). The data is accessed by functions that extract requested assays and their respective metadata, reconstruct the SCE(s) and return an MAE.



<br><br><br>
