---
title: "scMultiome tutorial"
output:
  BiocStyle::html_document:
    toc: true
    number_section: true
    self_contained: true
    titlecaps: true
vignette: >
  %\VignetteIndexEntry{scMultiome tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```


# Introduction

Single cell data is gaining sophistication - Cells can be measured in multiple modalities including gene expression, chromatin accessibility, cell surface markers and protein expression. These orthogonal measures of the same or matched cells enable a holistic construction of the cell state. However it has been challenging to share multiomic data, especially in an integrated format that consolidates the multiple layers of measurements. The `MultiAssayExperiment` provides a framework to package the various modalities into a single dataset on a per cell basis.

The `scMultiome` package is a collection of public single cell multiome data sets pre-processed and packaged into `MultiAssayExperiment` objects for downstream analysis. It also provides basic functions to save the `MultiAssayExperiment` as `.hdf5` files so that users can load only the desired modalities into memory. `scMultiome` provides interface with the package `ArchR` [package](https://www.archrproject.com/) by allowing the conversion of an ArchR project to a `MultiAssayExperiment`.


# Installation
``` {r, eval = FALSE}
if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")

BiocManager::install("scMultiome")

```


# Accessing datasets

## Available Data Sets

Use the `listDatasets` function to view currently available data sets and the names of their accessor functions. Help pages fot particular accessors contain more information on their data sets.
```{r, eval = FALSE}
library(scMultiome)
listDatasets()
```
```{r, echo = FALSE, results = 'asis'}
lds <- listDatasets()
knitr::kable(lds, caption = "Available Data Sets")

```

Access a data set by calling its accessor function:
```{r eval = FALSE}
prostateENZ()

```


See the help files for accessor functions for possible extra options, e.g. `?prostateENZ`.


## Transcription Factor Binding Motifs

The package contains additional data sets that contain information on transcription factor binding. These are not multiome data but compiled ChIP-seq data that can be used together with multiomes to compute epiregulons. Currently three genomic builds genomes are provided: hg38, hg19, and mm10.

The ChIP-seq data is packaged into individual RDS files but they are accessed with a common accessor function, `tfBinding`. Only one subset can be retrieved at a time, a genome must be specified.
```{r, eval = FALSE}
tfBinding("hg38")

```

<br><br><br>

# Processing data

Current multiomic data sets consist of gene expression and chromatin accessibility but can be extended to include any other modalities. The data sets are either paired multiomic data sets or unpaired data sets with data integration performed by the `ArchR` [package](https://www.archrproject.com/). The `ArchR` projects were converted to `MultiAssayExperiment` objects. [MultiAssayExperiment](https://www.bioconductor.org/packages/devel/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html) objects can be constructed easily from individual matrices pre-processed by users' favorite packages.

The `MultiAssayExperiment` object is saved in hdf5 files. The nature of the hdf5 format allows the MAEs to be split into individual experiments and store them in one file, so you can choose freely which ones to load. Experiments, usually `SingleCellExperiment` objects, are disassembled into parts, which are saved in the hdf5 hierarchy. Assays are saved as sparse arrays to save storage.

Upon loading, selected experiments are reassembled and wrapped into an MAE object. Assays are represented by `DelayedMatrix` objects to save memory.


## Raw Data Analysis

Raw data is first analyzed with the [10x Genomics Cell Ranger](https://www.10xgenomics.com/support) pipelines.

Following that, gene expression data is run through a [scran.chan](https://github.com/LTLA/scran.chan) analysis pipeline. Refer to the [OSCA book](https://bioconductor.org/books/release/OSCA/), specifically [this chapter](http://bioconductor.org/books/3.15/OSCA.basic/index.html) for details.


## ArchR Multiome Workflow

The general ArchR workflow is outlined below.
Refer to the [ArchR book](https://www.archrproject.com/bookdown/index.html) for details.

1. Fragment files resulting from Cell Ranger ATAC analysis are used to create arrow files. The _Tile Matrix_ and _Gene Score Matrix_ are created.
2. An ArchR project is created. Project-wide Tile and Gene Score Matrices are now available.
3. Duplex cells are filtered out. This step is omitted in cell culture based experiments.
4. Dimensionality reduction is run on the Tile Matrix using the Iterative LSI algorithm using `ArchR::addIterativeLSI`.
5. scRNAseq data is added to the ArchR project with `ArchR::addGeneExpressionMatrix` for paired datasets or with `ArchR::addGeneIntegrationMatrix` for unpaired ones. The _Gene Expression Matrix_ or the _Gene Integration Matrix_ is added.
6. Iterative LSI is run again on the _Gene Expression/Integration Matrix_.
7. Reduced dimension representations of scATAC and scRNAseq spaces are combined with `ArchR::addCombinedDims`. This creates a joint reduced dimensions representation that is used downstream.
8. Clustering is done with `ArchR::addClusters` on the iLSI representation of a space of the user's choosing: the scATAC space, the scRNAseq space, or the combined space..
9. UMAP and/or TSNE embeddings are created with `ArchR::addUMAP` and `ArchR::addTSNE`. Likewise, the user chooses the iLSI space to create embeddings from.
10. Pseudo-bulk replicates are computed with `ArchR::addGroupCoverages`.
11. Peaks are called with `ArchR::addReproduciblePeakSet`. (This requires `MACS2`.)
12. _Peak Matrix_ is added with `ArchR::addPeakMatrix`.
13. Peak annotation is added with `ArchR::addMotifAnnotations`. (This requires `chromVARmotifs`.)
14. Background peaks are obtained with `ArchR::addBgdPeaks`.
15. ChromVAR deviations are computed with `ArchR::addDeviationsMatrix`. The _Motif Matrix_ is added.


## Conversion to MAE

ArchR projects were converted to `MultiAssayExperiment` objects with `archr2MAE`. Each Matrix in the project becomes a `SingleCellExperiment`, its `colData` is derived from the project's `cellColData` slot, and `rowData` and `rowRanges` are created as needed. Reduced dimensionalities and embeddings of the project are attached to the matrices they were obtained from and are stored in the `reducedDims` slot.

```{r archr2MAE, eval = FALSE}
# convert ArchR project to MAE
mae <- archr2MAE("<ARCHR_PATH>")

```


## Contributing

If you want to contribute your publicly available multiome data set, please read the vignette _Adding Datasets_ and contact the package maintainer, `r desc::desc_get_maintainer()`.



<br><br><br>

# Adding Datasets

This package is a data repository for single cell multiome data within `ExperimentHub`. It also provides a framework to add more multiome data with minimal effort but the process cannot be fully automated. This vignette will guide you.

Briefly, data set will be disassembled and saved in a hdf5 file, from which individual experiments can be drawn using its own accessor function. You must provide the data set in hdf5 format, its metadata, its documentation, and finally the accessor function that will retrieve the data from `ExperimentHub`.


## Developer Mode

Adding data involves updating the package and as such it must be done in "developer mode". The developer mode allows access to additional tools such as documentation templates.

To work in developer mode, you must first clone the package repository with `git`. Create a branch from `master` and work on that.

Start an R session in the package directory (e.g. by opening the RStudio project in RStudio) and load all the functions. This is necessary for the R engine to temporarily identify your working directory as the package installation directory, and to expose the internal functions that you will be using.

```{r, eval = FALSE, devtoolsAvailable, include = FALSE}
devtoolsAvailable <- requireNamespace("devtools", quietly = TRUE)

if (devtoolsAvailable) {
    # attach development version of the package
    devtools::load_all()
}

```


## Saving Your Data Set

Currently only `MultiAssayExperiment` objects are supported. Experiments must be objects that inherit from `SummarizedExperiment` and will usually be `SingleCellExperiment`, hence full support is provided for the latter and their slots (`reducedDims` and `altExps`).

Use `saveMAE` to save your data set to the `R/data` directory. You can use `testFile` to validate that your data set can be reconstructed.

```{r saving, class.output = "scroll250", class.echo = "scroll250"}
# construct a dummy data set
mae <- dummyMAE()
mae

# name the file to save to
fileName <- tempfile(fileext = ".h5")

# save data set
saveMAE(mae, fileName)

# test reloading
testFile(fileName)

```

For a detailed explanation of the process see `?saveMAE`.


## Creating Metadata

Run `makeMakeMetadata("<DATASET_NAME>")` to create a script to store your data set's metadata. It will be called `inst/scripts/make-metadata-<DATASET_NAME>.R`. Metadata must be a 1-row data frame with specific columns and values must be character strings (some fields allow character vectors). See `inst/scripts/make-metadata.R` for more information and `inst/scripts/make-metadata-prostateENZ.R` for an example.

The file also stores metadata that will be returned by `listDatasets`. Likewise, this must be a 1-row data frame and values must be character strings.


```{r making metadata, eval = FALSE}
makeMakeMetadata("dataset")

```

Once your `make-metadata` file is ready, build the metadata by running the `inst/scripts/make-metadata.R` script. This must run without errors. If successful,
all the datasets will be captured in `inst/extdata/manifest.csv` and `inst/extdata/metadata.csv`

```{r build metadata}
# this will add your metadata
source(system.file("scripts", "make-metadata.R", package = "scMultiome"))

```

Subsequently, validate the correctness of the metadata by the running `makeExperimentHubMetadata` function. This call must also run without errors. It will return an `ExperimentHub` object that will display your metadata in the form that the end users will see it.


```{r validation, class.output = "scroll250"}
# this will validate your metadata (not shown)
ExperimentHubData::makeExperimentHubMetadata(dirname(system.file(package = "scMultiome")))

```


## Documenting Your Data Set

Every data set needs a help page to describe the source data, references, and whatever else you deem necessary.
Since every data set is accessed by its own accessor function, the `R` file that defines the function is the natural place to place the data set documentation. The package framework is constructed such that accessor functions are extremely simple and you can basically copy the original one (`prostateENZ`) and most of its documentation.

The process of obtaining the data set should also be described.

First, run `makeMakeData("<DATASET_NAME>")`. This will create an Rmarkdown file called `inst/scripts/make-data-<DATASET_NAME>.Rmd`. Use it to describe how your data set was obtained. This doesn't have to be a working script, just a report. Pseudocode is acceptable. Note that code evaluation has been disabled so that you can copy your actual code and the lengthy ArchR analysis does not run again.

```{r making, eval = FALSE}
makeMakeData("dataset")

```

Second, run `makeR("<DATASET_NAME>")`. This will create an R file called `<DATASET_NAME>.R`, which will provide access and help for your data set. This file will quote the accompanying Rmd file created above. This way the R file itself is more concise and easier to edit. Adjust the file accordingly:

1. Give the file the same Title was used in the `make-metadata-<>.R` file.
2. Add a Description section.
3. Document any arguments other than `experiments` and `metadata`.
4. Describe the format of your `MultiAssayExperiment`.
5. Cite appropriate references.
6. Make sure the default value of the `experiments` argument reflects the experiment names in your data set.
7. If you want to restore custom classes to your experiments, add converting functions here. They do not require documentation or exporting.


```{r making2, eval = FALSE}
makeR("dataset")

```

Once the files are ready, build the documentation, and `?scMultiome` and `?<DATASET_NAME>` to review it.
```{r documentation}
# build documentation
devtools::document()

# view package man page
?scMultiome

# view your data set man page
help("dataset")

```


## Updating Package Files

Edit the `DESCRIPTION` file to update package metadata: add yourself as package author and package version.

```{r DESCRIPTION, eval = FALSE}
desc::desc_add_author("<GIVEN_NAME>", "<FAMILY_NAME>", "<EMAIL>", role = "aut")
desc::desc_bump_version("minor")

```

Push your changes to git and create a merge request. Once the request is approved and merged to the `master` branch, the package maintainer will take over and move on to the next stage.


## Uploading Your Data Set

The final step of the process is to place the data set in Bioconductor's data bucket. This can only be done with Bioconductor's knowledge and blessing.

Bioconductor will be notified that a `scMultiome` update is coming by email at `hubs@bioconductor.org`. They will issue a temporary SAS token for the Bioconductor data bucket. The data set will be placed in the Bioconductor staging directory with `uploadFile` and Bioconductor will be notified again that the upload is ready. They will receive a link to the package repository, update metadata in `ExperimentHub` and finalize the process.

```{r upload, eval = FALSE}
uploadFile(file = fileName, sasToken = "<SAS_TOKEN>")

```

<br>

Consult [this vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/HubPub/inst/doc/CreateAHubPackage.html#additional-resources-to-existing-hub-package) for current Bioconductor requirements.


## Congratulations

If you found any of this vignette or the process confusing, we would welcome feedback and gladly add more clarifications. Please email the package maintainer, `r desc::desc_get_maintainer()`.


# Session Info

```{r}
sessionInfo()
```
<br><br><br>
